File: java/leetcode_grind/Day622.java
package leetcode_grind;

public class Day622 {
    // https://leetcode.com/problems/filling-bookcase-shelves/description/?envType=daily-question&envId=2024-07-31
    static class Solution1 {
        public int minHeightShelves(int[][] books, int shelfWidth) {
            int[][] memo = new int[books.length][shelfWidth + 1];
            return dpHelper(books, shelfWidth, memo, 0, shelfWidth, 0);
        }

        int dpHelper(int[][] books, int shelfWidth, int[][] memo, int i, int remainingShelfWidth, int maxHeight) {
            int[] currentBook = books[i];
            int maxHeightUpdated = Math.max(maxHeight, currentBook[1]);
            if (i == books.length - 1) {
                if (remainingShelfWidth >= currentBook[0]) {
                    return maxHeightUpdated;
                }
                return maxHeight + currentBook[1];
            }

            if (memo[i][remainingShelfWidth] != 0) {
                return memo[i][remainingShelfWidth];
            }

            int ans = maxHeight + dpHelper(
                    books, shelfWidth, memo, i + 1, shelfWidth - currentBook[0], currentBook[1]);

            if (remainingShelfWidth >= currentBook[0]) {
                ans = Math.min(ans, dpHelper(
                        books, shelfWidth, memo, i + 1, remainingShelfWidth - currentBook[0], maxHeightUpdated));
            }
            return memo[i][remainingShelfWidth] = ans;
        }
    }
}

File: src/leetcode_grind/day_622.rs
// https://leetcode.com/problems/filling-bookcase-shelves/description/?envType=daily-question&envId=2024-07-31
pub fn min_height_shelves(books: Vec<Vec<i32>>, shelf_width: i32) -> i32 {
    let n = books.len();
    let mut dp = vec![0; n + 1];
    dp[0] = 0;
    dp[1] = books[0][1];

    for i in 2..=n {
        let mut remaining_shelf_width = shelf_width - books[i - 1][0];
        let mut max_height = books[i - 1][1];
        dp[i] = books[i - 1][1] + dp[i - 1];

        let mut j = i - 1;

        while j > 0 && remaining_shelf_width - books[j - 1][0] >= 0 {
            max_height = max_height.max(books[j - 1][1]);
            remaining_shelf_width -= books[j - 1][0];
            dp[i] = dp[i].min(max_height + dp[j - 1]);
            j -= 1;
        }
    }
    dp[n]
}
